src/common/docker-compose/index.ts

```diff
@@ -444,4 +444,72 @@ export class DockerComposeUtils {
+    const max_restarts = 3;
+    const max_unhealthy_checks = 3;
+    const reset_timer = 60000;
+    let continueToRun = true;
+    const service_data_dictionary: Dictionary<{ restarts: number, last_restart_ms: number, unhealthy_count: number }> = {};
+    while (continueToRun) {
+      await new Promise(r => setTimeout(r, 5000));
+      const container_states = JSON.parse((await DockerComposeUtils.dockerCompose(['-f', compose_file, '-p', environment_name, 'ps', '--format', 'json'])).stdout);
+      for (const container_state of container_states) {
+        const id = container_state.ID;
+        const full_service_name = container_state.Service;
+        const service_name = container_state.Service.split('--')[1];
+        const state = container_state.State.toLowerCase();
+        const health = container_state.Health.toLowerCase();
+
+        if (!service_data_dictionary[service_name]) {
+          service_data_dictionary[service_name] = {
+            restarts: 0,
+            last_restart_ms: Date.now(),
+            unhealthy_count: 0,
+          };
+        }
+
+        let bad_state = state != 'running';
+        const bad_health = health == 'unhealthy';
+
+        // We do not want to call it a bad state just because it is unhealthy once
+        // So we make sure we are unhealthy for n iterations before calling it a bad state
+        if (bad_health) {
+          service_data_dictionary[service_name].unhealthy_count++;
+          const tries_left = Math.max(max_unhealthy_checks - service_data_dictionary[service_name].unhealthy_count, 0);
+          console.log(chalk.yellow(`Health check for ${service_name} has failed. ${tries_left} attempts left`));
+          if (service_data_dictionary[service_name].unhealthy_count >= max_unhealthy_checks) {
+            bad_state = true;
+            service_data_dictionary[service_name].unhealthy_count = 0;
+          }
+        } else {
+          service_data_dictionary[service_name].unhealthy_count = 0;
+        }
+
+        if (bad_state) {
+          const service_data = service_data_dictionary[service_name];
+
+          if (Date.now() - service_data.last_restart_ms > reset_timer) {
+            service_data.restarts = 0;
+          }
+
+          service_data.restarts++;
+          service_data.last_restart_ms = Date.now();
+          if (service_data.restarts > max_restarts) {
+            throw new Error(`Unable to start service ${service_name}`);
+          }
+          console.log(chalk.red(`ERROR: Service ${service_name} has encountered an error and is being restarted.`));
+          console.log(chalk.red(`Retry attempt ${service_data.restarts} of ${max_restarts}`));
+          await restart(id);
+          DockerComposeUtils.dockerCompose(['-f', compose_file, '-p', environment_name, 'logs', full_service_name, '--follow', '--since', new Date(service_data.last_restart_ms).toISOString()], { stdout: 'inherit' });
```

Comment:
![image](https://user-images.githubusercontent.com/737712/162525001-6bcb4d1d-bf54-426d-864e-9c16e0244701.png)

This seems to be creating multiple log streams for a service when it fails.

Reply:
I don't think this is necessary. I removed the line and my logs still show for the restarted containers.

Reply:
Yeah I removed and still got the api logs
![image](https://user-images.githubusercontent.com/737712/162526545-07bcc75c-c411-4d65-94d5-1d08e9bee595.png)

Reply:
Interesting for me the streams would be gone. Need to deep dive. Can I get your docker/kubernetes version number just to verify?

Reply:
```
Client: Docker Engine - Community
 Cloud integration: v1.0.22
 Version:           20.10.12
 API version:       1.41
 Go version:        go1.16.12
 Git commit:        e91ed57
 Built:             Mon Dec 13 11:45:48 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.12
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.12
  Git commit:       459d0df
  Built:            Mon Dec 13 11:43:56 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.12
  GitCommit:        7b11cfaabd73bb80907dd23182b9347b4245eb5d
 runc:
  Version:          1.0.2
  GitCommit:        v1.0.2-0-g52b36a2
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```